course/versioning.py

```diff
@@ -702,47 +704,42 @@ def unauthorized_access():
                     "utf-8", errors="replace")
         except binascii.Error:
             return unauthorized_access()
-        auth_data_values = auth_data.split(':', 1)
-        if len(auth_data_values) != 2:
-            return unauthorized_access()
-        username, token = auth_data_values
-        try:
-            possible_user = get_user_model().objects.get(
-                    username=username)
-        except ObjectDoesNotExist:
-            return unauthorized_access()
-
-        token_values = token.split('_', 1)
-        if len(token_values) != 2:
-            return unauthorized_access()
 
-        try:
-            int(token_values[0])
-        except ValueError:
+        match = GIT_AUTH_DATA_RE.match(auth_data)
+        if match is None:
             return unauthorized_access()
 
         from django.utils.timezone import now
-        from course.auth import find_matching_token
         now_datetime = now()
 
-        user_token = find_matching_token(course_identifier, token_values[0],
-                    token_values[1], now_datetime)
+        auth_data = dict(
+                course_identifier=course_identifier,
+                token_id=int(match.group(2)),
+                token_hash_str=match.group(3),
+                now_datetime=now_datetime)
+        username = match.group(1)
+
+        from course.auth import auth_course_with_token
+
+        try:
+            user, token = auth_course_with_token(request, **auth_data)
+        except PermissionDenied:
+            return unauthorized_access()
 
-        if user_token is None or user_token.user.username != username:
+        if user.username != username:
             return unauthorized_access()
 
-        user = possible_user
-        participation = user_token.participation
+        participation = token.participation
 
     if user is None:
         return unauthorized_access()
 
     course = participation.course
 
-    if user_token is not None and \
-            user_token.restrict_to_participation_role is not None:
-        check_permission = user_token.restrict_to_participation_role.has_permission
-        if course != user_token.restrict_to_participation_role.course:
+    if token is not None and \
+            token.restrict_to_participation_role is not None:
+        check_permission = token.restrict_to_participation_role.has_permission
+        if course != token.restrict_to_participation_role.course:
```

Comment:
Thanks! That said, I do think it's possible to unify these completely, so that the git endpoint can just use the decorator. What's needed (IMO) is to generalize the decorator to accept both the bearer token and basic auth as required by git.